### PR TITLE
Adding validation to branch creation

### LINKED ISSUE #20: Fix Create branch function to disallow branch that starts or end with '/'
If the branch name starts or ends with "/" it is invalid.

### LINKED ISSUE #67: Creating branch can fail silently on name conflict
The following actions fail silently from the user's perspective.

1. If branch `foobar` exists, attempting to create a branch named `foobar`
2. If branch `foo/bar` exists, attempting to create a branch named `foo`

```
170629-17:02:31 TRACE [23] <TaskManager>                       Exception on exclusive thread: 2 git branch
fatal: A branch named 'asdf' already exists.
  at GitHub.Unity.ProcessTask`1+<RunWithReturn>c__AnonStorey0[System.String].<>m__0 () [0x00116] in C:\Users\Spade\Projects\GitHub\UnityInternal\src\GitHub.Api\NewTaskSystem\ProcessTask.cs:295 
  at GitHub.Unity.ProcessWrapper.Run () [0x002fc] in C:\Users\Spade\Projects\GitHub\UnityInternal\src\GitHub.Api\NewTaskSystem\ProcessTask.cs:164 
  at GitHub.Unity.ProcessTask`1[T].RunWithReturn (Boolean success) [0x00063] in C:\Users\Spade\Projects\GitHub\UnityInternal\src\GitHub.Api\NewTaskSystem\ProcessTask.cs:305 
  at GitHub.Unity.TaskBase`1[TResult].<TaskBase>m__0 () [0x00043] in C:\Users\Spade\Projects\GitHub\UnityInternal\src\GitHub.Api\NewTaskSystem\TaskBase.cs:406 
  at System.Threading.Tasks.Task`1[System.String].InvokeFuture (System.Object futureAsObj) [0x00000] in <filename unknown>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x00000] in <filename unknown>:0 
  at System.Threading.Tasks.Task.Execute () [0x00000] in <filename unknown>:0 
   at GitHub.Unity.LogFacade.Trace(System.Exception ex, System.String message) in C:\Users\Spade\Projects\GitHub\UnityInternal\src\GitHub.Logging\LogFacade.cs:line 91
   at GitHub.Unity.TaskManager+<ScheduleExclusive>c__AnonStorey1`1[[GitHub.Unity.TaskBase, GitHub.Api, Version=0.12.0.0, Culture=neutral, PublicKeyToken=null]].<>m__0(System.Threading.Tasks.Task tt) in C:\Users\Spade\Projects\GitHub\UnityInternal\src\GitHub.Api\NewTaskSystem\TaskManager.cs:line 130
   at System.Threading.Tasks.Task+<>c__DisplayClass6.<ContinueWith>b__5(System.Object obj)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   at System.Threading.Tasks.Task.ExecutionContextCallback(System.Object obj)
   at System.Security.SecurityContext.Run(System.Security.SecurityContext securityContext, System.Threading.ContextCallback callback, System.Object state) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Security/SecurityContext.cs:line 150
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Threading/ExecutionContext.cs:line 150
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef currentTaskSlot)
   at System.Threading.Tasks.Task.ExecuteEntry(Boolean bPreventDoubleExecution)
   at System.Threading.Tasks.TaskScheduler.TryExecuteTask(System.Threading.Tasks.Task task)
   at GitHub.Unity.ConcurrentExclusiveInterleave+ConcurrentExclusiveTaskScheduler.ExecuteTask(System.Threading.Tasks.Task task) in C:\Users\Spade\Projects\GitHub\UnityInternal\src\GitHub.Api\NewTaskSystem\ConcurrentExclusiveInterleave.cs:line 398
   at GitHub.Unity.ConcurrentExclusiveInterleave.ExecuteConcurrentTask(System.Threading.Tasks.Task task) in C:\Users\Spade\Projects\GitHub\UnityInternal\src\GitHub.Api\NewTaskSystem\ConcurrentExclusiveInterleave.cs:line 214
   at System.Threading.Tasks.Parallel+<>c__DisplayClass32`2[[System.Threading.Tasks.Task, System.Threading, Version=1.0.2856.102, Culture=neutral, PublicKeyToken=31bf3856ad364e35],[System.Object, mscorlib, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<PartitionerForEachWorker>b__30()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.InnerInvokeWithArg(System.Threading.Tasks.Task childTask)
   at System.Threading.Tasks.Task+<>c__DisplayClass3.<ExecuteSelfReplicating>b__2(System.Object )
   at System.Threading.Tasks.Task.ExecuteSelfReplicating(System.Threading.Tasks.Task root)
   at System.Threading.Tasks.Task.ExecutionContextCallback(System.Object obj)
   at System.Security.SecurityContext.Run(System.Security.SecurityContext securityContext, System.Threading.ContextCallback callback, System.Object state) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Security/SecurityContext.cs:line 150
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Threading/ExecutionContext.cs:line 150
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef currentTaskSlot)
   at System.Threading.Tasks.Task.ExecuteEntry(Boolean bPreventDoubleExecution)
   at System.Reflection.MonoMethod.InternalInvoke(System.Object , System.Object[] , System.Exception ByRef )
   at System.Reflection.MonoMethod.Invoke(System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/MonoMethod.cs:line 222
   at System.Reflection.MethodBase.Invoke(System.Object obj, System.Object[] parameters) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/MethodBase.cs:line 115
   at GitHub.Unity.TaskSchedulerExcludingThread.ExecuteEntry(System.Threading.Tasks.Task task, Boolean flag) in C:\Users\Spade\Projects\GitHub\UnityInternal\src\GitHub.Api\NewTaskSystem\ConcurrentExclusiveInterleave.cs:line 33
   at GitHub.Unity.TaskSchedulerExcludingThread.TaskExecuteWaitCallback(System.Object obj) in C:\Users\Spade\Projects\GitHub\UnityInternal\src\GitHub.Api\NewTaskSystem\ConcurrentExclusiveInterleave.cs:line 49
```

### PR DESCRIPTION
**Pull request by StanleyGoldman**
Fixes #20 
Fixes #67 

Adding some simple validation

### CODE REVIEW

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/BranchesView.cs [hunk 1]
```diff
@@ -17,6 +18,7 @@ class BranchesView : Subview
         private const string NewBranchConfirmButton = "Create";
         private const string FavoritesSetting = "Favorites";
         private const string FavoritesTitle = "Favorites";
+        private const string CreateBranchTitle = "Create branch";
```
Comment (CapnRat) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/BranchesView.cs]: Should use title capitalization for labels. http://titlecapitalization.com/

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/BranchesView.cs [hunk 2]
```diff
@@ -498,7 +502,21 @@ private void OnCreateGUI()
...
+                                         Refresh();
+                                     }
+                                     else
+                                     {
+                                         var errorHeader = "fatal: ";
+                                         var errorMessage = e.Message.StartsWith(errorHeader) ? e.Message.Remove(0, errorHeader.Length) : e.Message;
+
+                                         EditorUtility.DisplayDialog(CreateBranchTitle,
+                                             errorMessage,
+                                             Localization.Cancel);
```
Comment (CapnRat) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/BranchesView.cs]: Alert dialogs that are purely informational should provide an OK button for acceptance. Cancel can be confusing/ambiguous.

https://developer.apple.com/macos/human-interface-guidelines/windows-and-views/alerts/

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/BranchesView.cs (hunk 1 above)
Comment (StanleyGoldman) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/BranchesView.cs]: 👍

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/BranchesView.cs (hunk 2 above)
Comment (StanleyGoldman) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/BranchesView.cs]: 👍

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/BranchesView.cs [hunk 3]
```diff
@@ -452,6 +454,8 @@ private void OnCreateGUI()
                     var cannotCreate = selectedNode == null ||
                                        selectedNode.Type == NodeType.Folder ||
                                        newBranchName == null ||
+                                       newBranchName.StartsWith("/") ||
+                                       newBranchName.EndsWith("/") ||
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/BranchesView.cs]: Oh, actually there's a bunch more validation we should be doing. Spaces need to be turned into `-`, and https://www.kernel.org/pub/software/scm/git/docs/git-check-ref-format.html has a bunch of rules:

- They can include slash / for hierarchical (directory) grouping, but no slash-separated component can begin with a dot . or end with the sequence .lock.
- They cannot have two consecutive dots .. anywhere.
- They cannot have ASCII control characters (i.e. bytes whose values are lower than \040, or \177 DEL), space, tilde ~, caret ^, or colon : anywhere.
- They cannot have question-mark ?, asterisk *, or open bracket [ anywhere.
- They cannot begin or end with a slash / or contain multiple consecutive slashes
- They cannot end with a dot ..
- They cannot contain a sequence @{.
- They cannot be the single character @.
- They cannot contain a \.

File: src/GitHub.Api/Helpers/RegularExpressions.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+namespace GitHub.Unity.Helpers
+{
+    public static class RegularExpressions
+    {
+        public static readonly Regex BranchNameRegex = new Regex(@"^(?<name>[\.\w\d\/\-_]+)$");
+    }
+
+    public static class BranchNameValidator
+    {
+        public static bool IsBranchNameValid(string branchName)
```
Comment (StanleyGoldman) [inline src/GitHub.Api/Helpers/RegularExpressions.cs]: Where else could/should this go? 🤷‍♂️

Comment (shana) [inline src/GitHub.Api/Helpers/RegularExpressions.cs]: Class name feels too specific. `Validation` would probably be good for a class name, then we can dump all the validator helpers in it

File: src/GitHub.Api/Helpers/RegularExpressions.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            return !string.IsNullOrEmpty(branchName)
+                && !branchName.Equals(".")
+                && !branchName.StartsWith("/")
+                && !branchName.EndsWith("/")
+                && !branchName.EndsWith(".")
+                && !branchName.Contains("//")
+                && !branchName.Contains(@"\")
+                && !branchName.EndsWith(".lock")
+                && !(branchName.StartsWith(".") && branchName.Contains("/"))
```
Comment (shana) [inline src/GitHub.Api/Helpers/RegularExpressions.cs]: `branchName.StartsWith("./")` perhaps?

Comment (shana) [inline src/GitHub.Api/Helpers/RegularExpressions.cs]: actually just `branchName.StartsWith(".")`, that's already not valid on the command line

File: src/GitHub.Api/Helpers/RegularExpressions.cs (hunk 4 above)
Comment (StanleyGoldman) [inline src/GitHub.Api/Helpers/RegularExpressions.cs]: Done.

File: src/GitHub.Api/Helpers/RegularExpressions.cs (hunk 5 above)
Comment (StanleyGoldman) [inline src/GitHub.Api/Helpers/RegularExpressions.cs]: And done.

**Pull request merged**